scripts/run_all.py

```diff
@@ -4,86 +4,239 @@
+
     process = subprocess.Popen(
         ["python3", "main.py", "--num-frames=30", "--steps=200"] + args,
         cwd=path,
     )
-    returncode = process.wait()
-    print(f"process exited with error code {returncode}")
+    if wait:
+        returncode = process.wait()
+        print(f"process exited with error code {returncode}")
+    return process
 
 
-def run_saved_example(path: str, args: List[str] = []) -> None:
+def run_saved_example(path: str, wait: bool = True) -> Any:
     process = subprocess.Popen(
-        ["cargo", "run", "-p", "rerun", "--all-features", "--", "out.rrd"] + args,
+        ["cargo", "run", "-p", "rerun", "--all-features", "--", "out.rrd"],
         cwd=path,
     )
-    returncode = process.wait()
-    print(f"process exited with error code {returncode}")
+    if wait:
+        returncode = process.wait()
+        print(f"process exited with error code {returncode}")
+    return process
 
 
-def collect_examples() -> List[str]:
-    return [os.path.dirname(entry) for entry in glob("examples/python/**/main.py")]
+def get_free_port() -> int:
+    with socket.socket() as s:
+        s.bind(("", 0))
+        return cast(int, s.getsockname()[1])
 
 
-def start_viewer(args: List[str] = []) -> Any:
-    process = subprocess.Popen(
-        ["cargo", "run", "-p", "rerun", "--all-features", "--"] + args,
-        stdout=subprocess.PIPE,
-    )
-    time.sleep(1)  # give it a moment to start
-    return process
+def collect_examples(fast: bool) -> List[str]:
+    if not fast:
+        return [os.path.dirname(entry) for entry in glob("examples/python/**/main.py")]
+    else:
+        # cherry picked
+        return [
+            "examples/python/deep_sdf",
+            "examples/python/dicom",
+            "examples/python/raw_mesh",
+            "examples/python/clock",
+            "examples/python/api_demo",
+            "examples/python/car",
+            "examples/python/colmap",
+            "examples/python/plots",
+            "examples/python/nyud",
+            "examples/python/text_logging",
+        ]
+
+
+class Viewer:
+    port: int
+    should_close: bool
+    web: bool
+    web_viewer_port: int
+    ws_server_port: int
+    process: Optional[Any]
+
+    def __init__(self, close: bool = False, web: bool = False):
+        self.port = get_free_port()
+        self.should_close = close
+        self.web = web
+        self.web_viewer_port = get_free_port()
+        self.ws_server_port = get_free_port()
+        self.process = None
+
+    def close(self) -> None:
+        if self.process is not None:
+            self.process.kill()
+        self.process = None
+
+    def start(self) -> "Viewer":
+        args = ["--port", str(self.port)]
+        if self.web:
+            args += [
+                "--web-viewer",
+                "--web-viewer-port",
+                str(self.web_viewer_port),
+                "--ws-server-port",
+                str(self.ws_server_port),
```

Comment:
nitpick: I think this should work and is a bit easier to read

```suggestion
                f"--web-viewer-port={self.web_viewer_port}",
                f"--ws-server-port={self.ws_server_port}",
```

Reply:
Agreed! Updated this, also updated other args to use fstrings.